android/src/main/java/co/uk/hive/reactnativegeolocation/location/LocationController.java

```diff
@@ -116,73 +113,43 @@ private boolean hasPermissions() {
     }
 
     private LocationRequest getLocationRequest(CurrentPositionRequest currentPositionRequest) {
-        return LocationRequest.create()
-                .setNumUpdates(1)
-                .setPriority(Priority.PRIORITY_BALANCED_POWER_ACCURACY)
-                .setExpirationDuration(currentPositionRequest.getTimeout());
+        return new LocationRequest.Builder(5000)
+                .setMaxUpdates(1)
+                .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
+                .setGranularity(Granularity.GRANULARITY_FINE)
+                .setDurationMillis(currentPositionRequest.getTimeout())
+                .build();
     }
 
     @MainThread
     private void requestLocation(CurrentPositionRequest currentPositionRequest,
                                  Function<LatLng, Object> successCallback,
                                  Function<Object, Object> failureCallback) {
-        final SingleLocationCallback singleLocationCallback = new SingleLocationCallback(successCallback, failureCallback);
-
-        final LocationCallback locationCallback = new LocationCallback() {
-            public void onLocationResult(LocationResult locationResult) {
-                mHandler.removeCallbacksAndMessages(null);
-                if (locationResult.getLastLocation() != null) {
-                    singleLocationCallback.locationReceived(new LatLng(
-                            locationResult.getLastLocation().getLatitude(),
-                            locationResult.getLastLocation().getLongitude()));
-                } else {
-                    singleLocationCallback.locationUnknown();
-                }
-            }
-        };
 
         if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                 && ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
             final String errorMessage = "Missing the required location permissions!";
             Log.e(LocationController.class.getName(), errorMessage);
-            failureCallback.apply(errorMessage);
+            failureCallback.apply(LocationError.PERMISSION_DENIED);
             return;
         }
-        mLocationClient.requestLocationUpdates(getLocationRequest(currentPositionRequest), locationCallback, Looper.getMainLooper());
```

Comment:
Requesting location updates was unreliable and wouldn't always complete. Assuming there was some throttling happening from the OS and the app was often stuck and timing out when trying to access the current location.